frontends/concrete-python/concrete/fhe/mlir/processors/assign_bit_widths.py

```diff
@@ -13,266 +13,346 @@
-    """Find the required precision of blocks and associate it corresponding nodes."""
-    size = len(nodes)
-    node_union = UnionFind(size)
-    for node_id, node in enumerate(nodes):
-        preds = graph.ordered_preds_of(node)
-        if not preds:
-            continue
-        # we always unify all inputs
-        first_input_id = preds[0].properties[NODE_ID]
-        for pred in preds[1:]:
-            pred_id = pred.properties[NODE_ID]
-            node_union.union(first_input_id, pred_id)
-        # we unify with outputs only if no precision change can occur
-        if not can_change_precision(node):
-            node_union.union(first_input_id, node_id)
-
-    blocks = convert_union_to_blocks(node_union)
-    result: list[None | tuple[Node, int]]
-    result = [None] * len(nodes)
-    for nodes_id in blocks:
-        output_p = required_encrypted_bitwidth(nodes[node_id] for node_id in nodes_id)
-        for node_id in nodes_id:
-            result[node_id] = (nodes[node_id], output_p)
-    assert None not in result
-    return typing.cast("list[tuple[Node, int]]", result)
-
-
-class UnionFind:
-    """
-    Utility class joins the nodes in equivalent precision classes.
-
-    Nodes are just integers id.
-    """
+            bit_width (z3.Int):
+                symbolic bit-width which will be assigned to node once constraints are solved
+        """
 
-    parent: list[int]
-
-    def __init__(self, size: int):
-        """Create a union find suitable for `size` nodes."""
-        self.parent = list(range(size))
-
-    @property
-    def size(self):
-        """Size in number of nodes."""
-        return len(self.parent)
-
-    def find_canonical(self, a: int) -> int:
-        """Find the current canonical node for a given input node."""
-        parent = self.parent[a]
-        if a == parent:
-            return a
-        canonical = self.find_canonical(parent)
-        self.parent[a] = canonical
-        return canonical
-
-    def union(self, a: int, b: int):
-        """Union both nodes."""
-        self.united_common_ancestor(a, b)
-
-    def united_common_ancestor(self, a: int, b: int) -> int:
-        """Deduce the common ancestor of both nodes after unification."""
-        parent_a = self.parent[a]
-        parent_b = self.parent[b]
-
-        if parent_a == parent_b:
-            return parent_a
-
-        if a == parent_a and parent_b < parent_a:
-            common_ancestor = parent_b
-        elif b == parent_b and parent_a < parent_b:
-            common_ancestor = parent_a
-        else:
-            common_ancestor = self.united_common_ancestor(parent_a, parent_b)
-
-        self.parent[a] = common_ancestor
-        self.parent[b] = common_ancestor
-        return common_ancestor
+        assert node.operation in {Operation.Generic, Operation.Constant, Operation.Input}
+        operation = (
+            node.properties["name"]
```

Comment:
not easy to read, I guess black fault.
```python
op = (
   bla if cond1
   else bli if cond2
   else blu
)
```
Locally disabling black or standard if ?

Reply:
I think it's okay, we can clearly see the condition, if, and else in one place. And else block is very self contained.